{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "ad53ebd2",
   "metadata": {},
   "source": [
    "# Expecto Model with multiple variants in input"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "07ddddf2",
   "metadata": {},
   "source": [
    "Working on the Expect model but with each input sequence having a combination of all possible SNPs inside the LD block"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6e0c043f",
   "metadata": {},
   "source": [
    "## Working to get sequences by building SNP class"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "910553dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Importing important modules\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import Bio\n",
    "import os\n",
    "from Bio import Entrez, SeqIO\n",
    "import itertools\n",
    "import argparse\n",
    "import math\n",
    "import torch\n",
    "from torch import nn\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import h5py\n",
    "from Multi_specto_class import *\n",
    "from Multi_specto_funcs import *\n",
    "Entrez.email  = \"pradluzog@gmail.com\"\n",
    "Entrez.api_key = \"98ad62666b4bd2dc831f1824727d74d67c08\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "50cf31d9",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Reading IGAP dataset\n",
    "\n",
    "igap = pd.read_csv('IGAP_stage_1.txt', sep='\\t')\n",
    "#for my mac\n",
    "#igap = pd.read_csv('/Users/pradeep/Downloads/IGAP_stage_1.txt', sep = '\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "c1443f49",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Position</th>\n",
       "      <th>MarkerName</th>\n",
       "      <th>Effect_allele</th>\n",
       "      <th>Non_Effect_allele</th>\n",
       "      <th>Beta</th>\n",
       "      <th>SE</th>\n",
       "      <th>Pvalue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6665078</th>\n",
       "      <td>19</td>\n",
       "      <td>45396665</td>\n",
       "      <td>rs59007384</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>0.9812</td>\n",
       "      <td>0.0208</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6665060</th>\n",
       "      <td>19</td>\n",
       "      <td>45388500</td>\n",
       "      <td>rs283811</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.9726</td>\n",
       "      <td>0.0218</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6665154</th>\n",
       "      <td>19</td>\n",
       "      <td>45427125</td>\n",
       "      <td>rs111789331</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>1.3811</td>\n",
       "      <td>0.0321</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6665058</th>\n",
       "      <td>19</td>\n",
       "      <td>45388130</td>\n",
       "      <td>rs34342646</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>1.1044</td>\n",
       "      <td>0.0246</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6665057</th>\n",
       "      <td>19</td>\n",
       "      <td>45387596</td>\n",
       "      <td>rs12972970</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>1.1072</td>\n",
       "      <td>0.0249</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6665056</th>\n",
       "      <td>19</td>\n",
       "      <td>45387459</td>\n",
       "      <td>rs12972156</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>1.1399</td>\n",
       "      <td>0.0256</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6665131</th>\n",
       "      <td>19</td>\n",
       "      <td>45415713</td>\n",
       "      <td>rs10414043</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>1.2958</td>\n",
       "      <td>0.0265</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6665112</th>\n",
       "      <td>19</td>\n",
       "      <td>45406673</td>\n",
       "      <td>rs10119</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.8220</td>\n",
       "      <td>0.0208</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6665148</th>\n",
       "      <td>19</td>\n",
       "      <td>45422946</td>\n",
       "      <td>rs41377151</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1.3511</td>\n",
       "      <td>0.0317</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         Chromosome  Position   MarkerName Effect_allele Non_Effect_allele  \\\n",
       "6665078          19  45396665   rs59007384             T                 G   \n",
       "6665060          19  45388500     rs283811             G                 A   \n",
       "6665154          19  45427125  rs111789331             A                 T   \n",
       "6665058          19  45388130   rs34342646             A                 G   \n",
       "6665057          19  45387596   rs12972970             A                 G   \n",
       "6665056          19  45387459   rs12972156             G                 C   \n",
       "6665131          19  45415713   rs10414043             A                 G   \n",
       "6665112          19  45406673      rs10119             A                 G   \n",
       "6665148          19  45422946   rs41377151             G                 A   \n",
       "\n",
       "           Beta      SE  Pvalue  \n",
       "6665078  0.9812  0.0208     0.0  \n",
       "6665060  0.9726  0.0218     0.0  \n",
       "6665154  1.3811  0.0321     0.0  \n",
       "6665058  1.1044  0.0246     0.0  \n",
       "6665057  1.1072  0.0249     0.0  \n",
       "6665056  1.1399  0.0256     0.0  \n",
       "6665131  1.2958  0.0265     0.0  \n",
       "6665112  0.8220  0.0208     0.0  \n",
       "6665148  1.3511  0.0317     0.0  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Filtering the igap snps \n",
    "igap = igap.sort_values(by = ['Pvalue'], ascending=True)\n",
    "top_10_igap_snps = igap.iloc[1:10,:]\n",
    "top_10_igap_snps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "273edbc1",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Defining a SNP class to perform simple LD filtering duties\n",
    "class SNP:\n",
    "    \n",
    "    def __init__(self,rsid,position,chromosome):\n",
    "        self.rsid = rsid\n",
    "        self.position = position\n",
    "        self.chr = chromosome\n",
    "    \n",
    "\n",
    "        \n",
    "    def check_ld_snps(self,dataset,window = 1000):\n",
    "        start_position = self.position - window + 1\n",
    "        end_position = self.position + window\n",
    "        dataset = dataset[dataset['Chromosome'] == self.chr]\n",
    "        def extract_neighbour_snps(start_position, end_position, dataset):\n",
    "            neighbour_snps = []\n",
    "            for index,row in dataset.iterrows():\n",
    "                if start_position <= dataset.loc[index,'Position'] <= end_position:\n",
    "                    neighbour_snps.append(dataset.loc[index,'MarkerName'])\n",
    "                else:\n",
    "                    continue\n",
    "            return neighbour_snps\n",
    "    \n",
    "        self.snps_in_window = extract_neighbour_snps(start_position,end_position,dataset)\n",
    "        return self.snps_in_window\n",
    "    \n",
    "    def obtain_snp_sequence(self,window = 1000):\n",
    "        start_position = self.position - window +1\n",
    "        end_position = self.position + window\n",
    "        if int(self.chr) < 10:\n",
    "            id_chr = \"\".join([\"NC_00000\",str(self.chr)])\n",
    "        else:\n",
    "            id_chr = \"\".join([\"NC_0000\",str(self.chr)])\n",
    "\n",
    "        handle = Entrez.efetch(db=\"nucleotide\",\n",
    "                        id = id_chr,\n",
    "                        rettype = \"fasta\",\n",
    "                        strand = 1,\n",
    "                        seq_start = start_position,\n",
    "                        seq_stop  = end_position)\n",
    "        record = SeqIO.read(handle,\"fasta\")\n",
    "        self.snp_sequence = str(record.seq)\n",
    "        return self.snp_sequence\n",
    "    \n",
    "    def obtain_all_comb_seq(self,dataset, window = 1000):\n",
    "        \n",
    "        def all_snp_combinations(a):\n",
    "            combinations = []\n",
    "            for k in range(0,len(a)):\n",
    "                t = list(itertools.combinations(a,k+1))\n",
    "                combinations.extend(t)\n",
    "            return combinations\n",
    "        \n",
    "        self.combinations = all_snp_combinations(self.snps_in_window)\n",
    "        comb_names = ['_'.join(x) for x in self.combinations if len(x)> 0]\n",
    "        comb_names.append('_'.join(['Ref',self.rsid]))\n",
    "        combination_dataset = dataset[dataset['MarkerName'].isin(self.snps_in_window)]\n",
    "        sequences = []\n",
    "        \n",
    "        for comb in self.combinations:\n",
    "            seq_to_change = self.snp_sequence\n",
    "            start_position = self.position - window + 1\n",
    "            end_position = self.position + window\n",
    "            for k in range(0,len(comb)):\n",
    "                idx = combination_dataset['MarkerName'] == comb[k]\n",
    "                pos = combination_dataset.loc[idx,'Position']\n",
    "                allele = str(combination_dataset.loc[idx,'Non_Effect_allele'].values[0])\n",
    "                net_pos = int(pos) - int(start_position)\n",
    "                seq_to_change = seq_to_change[:net_pos-1] + allele + seq_to_change[net_pos:]\n",
    "            sequences.append(seq_to_change)\n",
    "        sequences.append(self.snp_sequence)\n",
    "        sequences_named = dict(zip(comb_names,sequences))\n",
    "        return sequences_named\n",
    "                \n",
    "                \n",
    "    def seq_combination(self,dataset,window = 1000):\n",
    "        self.check_ld_snps(dataset,window)\n",
    "        self.obtain_snp_sequence()\n",
    "        self.combination_seq = self.obtain_all_comb_seq(dataset,window)\n",
    "        return self.combination_seq\n",
    "        \n",
    "    \n",
    "    def __str__(self):\n",
    "        return \"The SNP in object is \"+self.rsid\n",
    "        \n",
    "        \n",
    "        \n",
    "        \n",
    "        "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2f1dd7a2",
   "metadata": {},
   "source": [
    "## Remodelling Chromatin.py from Expecto"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "bd85b1ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Important library calls\n",
    "import argparse\n",
    "import math\n",
    "import torch\n",
    "from torch import nn\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import h5py"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "f46ba8b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Inputing the resources for Expect.py\n",
    "inputsize = 2000\n",
    "batchSize = 32\n",
    "maxshift = 800\n",
    "args_cuda = False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "fa2029c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "#DL model\n",
    "class LambdaBase(nn.Sequential):\n",
    "    def __init__(self, fn, *args):\n",
    "        super(LambdaBase, self).__init__(*args)\n",
    "        self.lambda_func = fn\n",
    "\n",
    "    def forward_prepare(self, input):\n",
    "        output = []\n",
    "        for module in self._modules.values():\n",
    "            output.append(module(input))\n",
    "        return output if output else input\n",
    "\n",
    "class Lambda(LambdaBase):\n",
    "    def forward(self, input):\n",
    "        return self.lambda_func(self.forward_prepare(input))\n",
    "\n",
    "class Beluga(nn.Module):\n",
    "    def __init__(self):\n",
    "        super(Beluga, self).__init__()\n",
    "        self.model = nn.Sequential(\n",
    "            nn.Sequential(\n",
    "                nn.Conv2d(4,320,(1, 8)),\n",
    "                nn.ReLU(),\n",
    "                nn.Conv2d(320,320,(1, 8)),\n",
    "                nn.ReLU(),\n",
    "                nn.Dropout(0.2),\n",
    "                nn.MaxPool2d((1, 4),(1, 4)),\n",
    "                nn.Conv2d(320,480,(1, 8)),\n",
    "                nn.ReLU(),\n",
    "                nn.Conv2d(480,480,(1, 8)),\n",
    "                nn.ReLU(),\n",
    "                nn.Dropout(0.2),\n",
    "                nn.MaxPool2d((1, 4),(1, 4)),\n",
    "                nn.Conv2d(480,640,(1, 8)),\n",
    "                nn.ReLU(),\n",
    "                nn.Conv2d(640,640,(1, 8)),\n",
    "                nn.ReLU(),\n",
    "            ),\n",
    "            nn.Sequential(\n",
    "                nn.Dropout(0.5),\n",
    "                Lambda(lambda x: x.view(x.size(0),-1)),\n",
    "                nn.Sequential(Lambda(lambda x: x.view(1,-1) if 1==len(x.size()) else x ),nn.Linear(67840,2003)),\n",
    "                nn.ReLU(),\n",
    "                nn.Sequential(Lambda(lambda x: x.view(1,-1) if 1==len(x.size()) else x ),nn.Linear(2003,2002)),\n",
    "            ),\n",
    "            nn.Sigmoid(),\n",
    "        )\n",
    "\n",
    "    def forward(self, x):\n",
    "        return self.model(x)\n",
    "\n",
    "\n",
    "\n",
    "def encodeSeqs(seqs, inputsize=2000):\n",
    "    \"\"\"Convert sequences to 0-1 encoding and truncate to the input size.\n",
    "    The output concatenates the forward and reverse complement sequence\n",
    "    encodings.\n",
    "    Args:\n",
    "        seqs: list of sequences (e.g. produced by fetchSeqs)\n",
    "        inputsize: the number of basepairs to encode in the output\n",
    "    Returns:\n",
    "        numpy array of dimension: (2 x number of sequence) x 4 x inputsize\n",
    "    2 x number of sequence because of the concatenation of forward and reverse\n",
    "    complement sequences.\n",
    "    \"\"\"\n",
    "    seqsnp = np.zeros((len(seqs), 4, inputsize), np.bool_)\n",
    "\n",
    "    mydict = {'A': np.asarray([1, 0, 0, 0]), 'G': np.asarray([0, 1, 0, 0]),\n",
    "            'C': np.asarray([0, 0, 1, 0]), 'T': np.asarray([0, 0, 0, 1]),\n",
    "            'N': np.asarray([0, 0, 0, 0]), 'H': np.asarray([0, 0, 0, 0]),\n",
    "            'a': np.asarray([1, 0, 0, 0]), 'g': np.asarray([0, 1, 0, 0]),\n",
    "            'c': np.asarray([0, 0, 1, 0]), 't': np.asarray([0, 0, 0, 1]),\n",
    "            'n': np.asarray([0, 0, 0, 0]), '-': np.asarray([0, 0, 0, 0])}\n",
    "\n",
    "    n = 0\n",
    "    for line in seqs:\n",
    "        cline = line[int(math.floor(((len(line) - inputsize) / 2.0))):int(math.floor(len(line) - (len(line) - inputsize) / 2.0))]\n",
    "        for i, c in enumerate(cline):\n",
    "            seqsnp[n, :, i] = mydict[c]\n",
    "        n = n + 1\n",
    "\n",
    "    # get the complementary sequences\n",
    "    dataflip = seqsnp[:, ::-1, ::-1]\n",
    "    seqsnp = np.concatenate([seqsnp, dataflip], axis=0)\n",
    "    return seqsnp\n",
    "\n",
    "def get_predicted_diff(snp_comb_seq,inputsize = 2000, batchSize = 32, maxshift = 800, args_cuda = False):\n",
    "    \"\"\"\n",
    "    Function to obtain all the predicted chromatin values for reference and alterante \n",
    "    and find the difference among them for further analysis.\n",
    "    Args:\n",
    "        snp_comb_seq: A dictionary of sequences as string object with A,T,G,C characters\n",
    "                        and keys corresponding to snps and combinations of snps with atleast\n",
    "                        one snp having 'Ref' in the key name to denote reference variant\n",
    "    Return:\n",
    "            A dictionary of matrix size 4000x2002 for the chromatin difference values for each \n",
    "            variant and combination except the reference\n",
    "    \"\"\"\n",
    "    refseqs = [seq for key, seq in snp_comb_seq.items() if 'ref' in key.lower()]\n",
    "    refseqs = refseqs[0]\n",
    "    ref_encoded = encodeSeqs(refseqs, inputsize=inputsize).astype(np.float32)\n",
    "\n",
    "    ref_preds = []\n",
    "    for i in range(int(1 + (ref_encoded.shape[0]-1) / batchSize)):\n",
    "        input = torch.from_numpy(ref_encoded[int(i*batchSize):int((i+1)*batchSize),:,:]).unsqueeze(2)\n",
    "        if args_cuda:\n",
    "            input = input.cuda()\n",
    "        ref_preds.append(model.forward(input).cpu().detach().numpy().copy())\n",
    "    ref_preds = np.vstack(ref_preds)\n",
    "    \n",
    "    comb_diff_pred = {}\n",
    "    for comb_seq in snp_comb_seq.keys():\n",
    "\n",
    "        if('Ref' not in comb_seq):\n",
    "\n",
    "            altseqs = snp_comb_seq[comb_seq]\n",
    "            alt_encoded = encodeSeqs(altseqs, inputsize=inputsize).astype(np.float32)\n",
    "\n",
    "            alt_preds = []\n",
    "            for i in range(int(1 + (alt_encoded.shape[0]-1) / batchSize)):\n",
    "                input = torch.from_numpy(alt_encoded[int(i*batchSize):int((i+1)*batchSize),:,:]).unsqueeze(2)\n",
    "                if args_cuda:\n",
    "                    input = input.cuda()\n",
    "                alt_preds.append(model.forward(input).cpu().detach().numpy().copy())\n",
    "            alt_preds = np.vstack(alt_preds)\n",
    "\n",
    "            diff = alt_preds - ref_preds\n",
    "            comb_diff_pred[comb_seq] = diff\n",
    "    \n",
    "    \n",
    "    return comb_diff_pred"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "71e90b4f",
   "metadata": {},
   "outputs": [],
   "source": [
    "model = Beluga()\n",
    "model.load_state_dict(torch.load('deepsea.beluga.pth'))\n",
    "model.eval()\n",
    "#model.cuda()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "2291e519",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "comb_diff_pred = get_predicted_diff(snp_comb_seq)\n",
    "f = h5py.File(snp_test +'.diff.h5', 'w')\n",
    "key_names = list(comb_diff_pred.keys())\n",
    "for i in key_names:\n",
    "    f.create_dataset(i, data=comb_diff_pred[i])\n",
    "f.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f1e58481",
   "metadata": {},
   "source": [
    "## Running chromatin prediction for top 10 SNPs and its combinations with LD blocks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "id": "4a909a6d",
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-93-46f072295b69>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[0;32m      3\u001b[0m     \u001b[0msnp_obj\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mSNP\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mtop_10_igap_snps\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0miloc\u001b[0m\u001b[1;33m[\u001b[0m\u001b[0mk\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;36m2\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m,\u001b[0m\u001b[0mtop_10_igap_snps\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0miloc\u001b[0m\u001b[1;33m[\u001b[0m\u001b[0mk\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;36m1\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m,\u001b[0m\u001b[0mtop_10_igap_snps\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0miloc\u001b[0m\u001b[1;33m[\u001b[0m\u001b[0mk\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;36m0\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      4\u001b[0m     \u001b[0msnp_comb_seq\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0msnp_obj\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mseq_combination\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0migap\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m----> 5\u001b[1;33m     \u001b[0mcomb_diff_pred\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mget_predicted_diff\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0msnp_comb_seq\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m      6\u001b[0m     \u001b[0mf\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mh5py\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mFile\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0msnp_test\u001b[0m \u001b[1;33m+\u001b[0m\u001b[1;34m'.diff.h5'\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;34m'w'\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      7\u001b[0m     \u001b[0mkey_names\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mlist\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mcomb_diff_pred\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mkeys\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m<ipython-input-9-3d655f6e5b62>\u001b[0m in \u001b[0;36mget_predicted_diff\u001b[1;34m(snp_comb_seq, inputsize, batchSize, maxshift, args_cuda)\u001b[0m\n\u001b[0;32m    122\u001b[0m                 \u001b[1;32mif\u001b[0m \u001b[0margs_cuda\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    123\u001b[0m                     \u001b[0minput\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0minput\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mcuda\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 124\u001b[1;33m                 \u001b[0malt_preds\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mmodel\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mforward\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0minput\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mcpu\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mdetach\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mnumpy\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mcopy\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    125\u001b[0m             \u001b[0malt_preds\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mnp\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mvstack\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0malt_preds\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    126\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m<ipython-input-9-3d655f6e5b62>\u001b[0m in \u001b[0;36mforward\u001b[1;34m(self, x)\u001b[0m\n\u001b[0;32m     48\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     49\u001b[0m     \u001b[1;32mdef\u001b[0m \u001b[0mforward\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mx\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 50\u001b[1;33m         \u001b[1;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mmodel\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mx\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     51\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     52\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32mC:\\ProgramData\\Anaconda3\\lib\\site-packages\\torch\\nn\\modules\\module.py\u001b[0m in \u001b[0;36m_call_impl\u001b[1;34m(self, *input, **kwargs)\u001b[0m\n\u001b[0;32m   1100\u001b[0m         if not (self._backward_hooks or self._forward_hooks or self._forward_pre_hooks or _global_backward_hooks\n\u001b[0;32m   1101\u001b[0m                 or _global_forward_hooks or _global_forward_pre_hooks):\n\u001b[1;32m-> 1102\u001b[1;33m             \u001b[1;32mreturn\u001b[0m \u001b[0mforward_call\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m*\u001b[0m\u001b[0minput\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;33m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m   1103\u001b[0m         \u001b[1;31m# Do not call functions when jit is used\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   1104\u001b[0m         \u001b[0mfull_backward_hooks\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mnon_full_backward_hooks\u001b[0m \u001b[1;33m=\u001b[0m \u001b[1;33m[\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;33m[\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32mC:\\ProgramData\\Anaconda3\\lib\\site-packages\\torch\\nn\\modules\\container.py\u001b[0m in \u001b[0;36mforward\u001b[1;34m(self, input)\u001b[0m\n\u001b[0;32m    139\u001b[0m     \u001b[1;32mdef\u001b[0m \u001b[0mforward\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0minput\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    140\u001b[0m         \u001b[1;32mfor\u001b[0m \u001b[0mmodule\u001b[0m \u001b[1;32min\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 141\u001b[1;33m             \u001b[0minput\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mmodule\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0minput\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    142\u001b[0m         \u001b[1;32mreturn\u001b[0m \u001b[0minput\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    143\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32mC:\\ProgramData\\Anaconda3\\lib\\site-packages\\torch\\nn\\modules\\module.py\u001b[0m in \u001b[0;36m_call_impl\u001b[1;34m(self, *input, **kwargs)\u001b[0m\n\u001b[0;32m   1100\u001b[0m         if not (self._backward_hooks or self._forward_hooks or self._forward_pre_hooks or _global_backward_hooks\n\u001b[0;32m   1101\u001b[0m                 or _global_forward_hooks or _global_forward_pre_hooks):\n\u001b[1;32m-> 1102\u001b[1;33m             \u001b[1;32mreturn\u001b[0m \u001b[0mforward_call\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m*\u001b[0m\u001b[0minput\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;33m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m   1103\u001b[0m         \u001b[1;31m# Do not call functions when jit is used\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   1104\u001b[0m         \u001b[0mfull_backward_hooks\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mnon_full_backward_hooks\u001b[0m \u001b[1;33m=\u001b[0m \u001b[1;33m[\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;33m[\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32mC:\\ProgramData\\Anaconda3\\lib\\site-packages\\torch\\nn\\modules\\container.py\u001b[0m in \u001b[0;36mforward\u001b[1;34m(self, input)\u001b[0m\n\u001b[0;32m    139\u001b[0m     \u001b[1;32mdef\u001b[0m \u001b[0mforward\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0minput\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    140\u001b[0m         \u001b[1;32mfor\u001b[0m \u001b[0mmodule\u001b[0m \u001b[1;32min\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 141\u001b[1;33m             \u001b[0minput\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mmodule\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0minput\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    142\u001b[0m         \u001b[1;32mreturn\u001b[0m \u001b[0minput\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    143\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32mC:\\ProgramData\\Anaconda3\\lib\\site-packages\\torch\\nn\\modules\\module.py\u001b[0m in \u001b[0;36m_call_impl\u001b[1;34m(self, *input, **kwargs)\u001b[0m\n\u001b[0;32m   1100\u001b[0m         if not (self._backward_hooks or self._forward_hooks or self._forward_pre_hooks or _global_backward_hooks\n\u001b[0;32m   1101\u001b[0m                 or _global_forward_hooks or _global_forward_pre_hooks):\n\u001b[1;32m-> 1102\u001b[1;33m             \u001b[1;32mreturn\u001b[0m \u001b[0mforward_call\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m*\u001b[0m\u001b[0minput\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;33m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m   1103\u001b[0m         \u001b[1;31m# Do not call functions when jit is used\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   1104\u001b[0m         \u001b[0mfull_backward_hooks\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mnon_full_backward_hooks\u001b[0m \u001b[1;33m=\u001b[0m \u001b[1;33m[\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;33m[\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32mC:\\ProgramData\\Anaconda3\\lib\\site-packages\\torch\\nn\\modules\\conv.py\u001b[0m in \u001b[0;36mforward\u001b[1;34m(self, input)\u001b[0m\n\u001b[0;32m    444\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    445\u001b[0m     \u001b[1;32mdef\u001b[0m \u001b[0mforward\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0minput\u001b[0m\u001b[1;33m:\u001b[0m \u001b[0mTensor\u001b[0m\u001b[1;33m)\u001b[0m \u001b[1;33m->\u001b[0m \u001b[0mTensor\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 446\u001b[1;33m         \u001b[1;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_conv_forward\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0minput\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mweight\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mbias\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    447\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    448\u001b[0m \u001b[1;32mclass\u001b[0m \u001b[0mConv3d\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0m_ConvNd\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32mC:\\ProgramData\\Anaconda3\\lib\\site-packages\\torch\\nn\\modules\\conv.py\u001b[0m in \u001b[0;36m_conv_forward\u001b[1;34m(self, input, weight, bias)\u001b[0m\n\u001b[0;32m    440\u001b[0m                             \u001b[0mweight\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mbias\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mstride\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    441\u001b[0m                             _pair(0), self.dilation, self.groups)\n\u001b[1;32m--> 442\u001b[1;33m         return F.conv2d(input, weight, bias, self.stride,\n\u001b[0m\u001b[0;32m    443\u001b[0m                         self.padding, self.dilation, self.groups)\n\u001b[0;32m    444\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "for k in range(0,len(top_10_igap_snps)):\n",
    "    snp_test = top_10_igap_snps.iloc[k,2]\n",
    "    snp_obj = SNP(top_10_igap_snps.iloc[k,2],top_10_igap_snps.iloc[k,1],top_10_igap_snps.iloc[k,0])\n",
    "    snp_comb_seq = snp_obj.seq_combination(igap)\n",
    "    comb_diff_pred = get_predicted_diff(snp_comb_seq)\n",
    "    f = h5py.File(snp_test +'.diff.h5', 'w')\n",
    "    key_names = list(comb_diff_pred.keys())\n",
    "    for i in key_names:\n",
    "        f.create_dataset(i, data=comb_diff_pred[i])\n",
    "    f.close()  \n",
    "    \n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
