{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "ad53ebd2",
   "metadata": {},
   "source": [
    "# Expecto Model with multiple variants in input"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "07ddddf2",
   "metadata": {},
   "source": [
    "Working on the Expect model but with each input sequence having a combination of all possible SNPs inside the LD block"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6e0c043f",
   "metadata": {},
   "source": [
    "## Working to get sequences by building SNP class"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "910553dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Importing important modules\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import Bio\n",
    "import os\n",
    "from Bio import Entrez, SeqIO\n",
    "import itertools\n",
    "import argparse\n",
    "import math\n",
    "from Bio import Entrez\n",
    "import xmltodict\n",
    "from pprint import pprint\n",
    "import torch\n",
    "from torch import nn\n",
    "import h5py\n",
    "from Multi_specto_class import *\n",
    "from Multi_specto_funcs import *\n",
    "Entrez.email  = \"pradluzog@gmail.com\"\n",
    "Entrez.api_key = \"98ad62666b4bd2dc831f1824727d74d67c08\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "50cf31d9",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Reading IGAP dataset\n",
    "\n",
    "igap = pd.read_csv('IGAP_stage_1.txt', sep='\\t')\n",
    "#for my mac\n",
    "#igap = pd.read_csv('~/Downloads/IGAP_stage_1.txt', sep = '\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "c1443f49",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Position</th>\n",
       "      <th>MarkerName</th>\n",
       "      <th>Effect_allele</th>\n",
       "      <th>Non_Effect_allele</th>\n",
       "      <th>Beta</th>\n",
       "      <th>SE</th>\n",
       "      <th>Pvalue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6665069</th>\n",
       "      <td>19</td>\n",
       "      <td>45394969</td>\n",
       "      <td>rs184017</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>0.9704</td>\n",
       "      <td>0.0208</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6665078</th>\n",
       "      <td>19</td>\n",
       "      <td>45396665</td>\n",
       "      <td>rs59007384</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>0.9812</td>\n",
       "      <td>0.0208</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6665060</th>\n",
       "      <td>19</td>\n",
       "      <td>45388500</td>\n",
       "      <td>rs283811</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.9726</td>\n",
       "      <td>0.0218</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6665154</th>\n",
       "      <td>19</td>\n",
       "      <td>45427125</td>\n",
       "      <td>rs111789331</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>1.3811</td>\n",
       "      <td>0.0321</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6665058</th>\n",
       "      <td>19</td>\n",
       "      <td>45388130</td>\n",
       "      <td>rs34342646</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>1.1044</td>\n",
       "      <td>0.0246</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6665057</th>\n",
       "      <td>19</td>\n",
       "      <td>45387596</td>\n",
       "      <td>rs12972970</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>1.1072</td>\n",
       "      <td>0.0249</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6665056</th>\n",
       "      <td>19</td>\n",
       "      <td>45387459</td>\n",
       "      <td>rs12972156</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>1.1399</td>\n",
       "      <td>0.0256</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6665131</th>\n",
       "      <td>19</td>\n",
       "      <td>45415713</td>\n",
       "      <td>rs10414043</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>1.2958</td>\n",
       "      <td>0.0265</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6665112</th>\n",
       "      <td>19</td>\n",
       "      <td>45406673</td>\n",
       "      <td>rs10119</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.8220</td>\n",
       "      <td>0.0208</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6665148</th>\n",
       "      <td>19</td>\n",
       "      <td>45422946</td>\n",
       "      <td>rs41377151</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1.3511</td>\n",
       "      <td>0.0317</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6665119</th>\n",
       "      <td>19</td>\n",
       "      <td>45410002</td>\n",
       "      <td>rs769449</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>1.2585</td>\n",
       "      <td>0.0257</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6665147</th>\n",
       "      <td>19</td>\n",
       "      <td>45422846</td>\n",
       "      <td>rs56131196</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>1.2895</td>\n",
       "      <td>0.0283</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6665146</th>\n",
       "      <td>19</td>\n",
       "      <td>45422160</td>\n",
       "      <td>rs12721051</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>1.3043</td>\n",
       "      <td>0.0278</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6665121</th>\n",
       "      <td>19</td>\n",
       "      <td>45411941</td>\n",
       "      <td>rs429358</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1.3503</td>\n",
       "      <td>0.0272</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6665142</th>\n",
       "      <td>19</td>\n",
       "      <td>45421254</td>\n",
       "      <td>rs12721046</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>1.2498</td>\n",
       "      <td>0.0284</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6665138</th>\n",
       "      <td>19</td>\n",
       "      <td>45418790</td>\n",
       "      <td>rs5117</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1.2736</td>\n",
       "      <td>0.0302</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6665125</th>\n",
       "      <td>19</td>\n",
       "      <td>45413576</td>\n",
       "      <td>rs75627662</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.8790</td>\n",
       "      <td>0.0228</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6665133</th>\n",
       "      <td>19</td>\n",
       "      <td>45416178</td>\n",
       "      <td>rs483082</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>0.9160</td>\n",
       "      <td>0.0222</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6665077</th>\n",
       "      <td>19</td>\n",
       "      <td>45396219</td>\n",
       "      <td>rs157582</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.9457</td>\n",
       "      <td>0.0212</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6665076</th>\n",
       "      <td>19</td>\n",
       "      <td>45396144</td>\n",
       "      <td>rs11556505</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>1.0651</td>\n",
       "      <td>0.0231</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         Chromosome  Position   MarkerName Effect_allele Non_Effect_allele  \\\n",
       "6665069          19  45394969     rs184017             G                 T   \n",
       "6665078          19  45396665   rs59007384             T                 G   \n",
       "6665060          19  45388500     rs283811             G                 A   \n",
       "6665154          19  45427125  rs111789331             A                 T   \n",
       "6665058          19  45388130   rs34342646             A                 G   \n",
       "6665057          19  45387596   rs12972970             A                 G   \n",
       "6665056          19  45387459   rs12972156             G                 C   \n",
       "6665131          19  45415713   rs10414043             A                 G   \n",
       "6665112          19  45406673      rs10119             A                 G   \n",
       "6665148          19  45422946   rs41377151             G                 A   \n",
       "6665119          19  45410002     rs769449             A                 G   \n",
       "6665147          19  45422846   rs56131196             A                 G   \n",
       "6665146          19  45422160   rs12721051             G                 C   \n",
       "6665121          19  45411941     rs429358             C                 T   \n",
       "6665142          19  45421254   rs12721046             A                 G   \n",
       "6665138          19  45418790       rs5117             C                 T   \n",
       "6665125          19  45413576   rs75627662             T                 C   \n",
       "6665133          19  45416178     rs483082             T                 G   \n",
       "6665077          19  45396219     rs157582             T                 C   \n",
       "6665076          19  45396144   rs11556505             T                 C   \n",
       "\n",
       "           Beta      SE  Pvalue  \n",
       "6665069  0.9704  0.0208     0.0  \n",
       "6665078  0.9812  0.0208     0.0  \n",
       "6665060  0.9726  0.0218     0.0  \n",
       "6665154  1.3811  0.0321     0.0  \n",
       "6665058  1.1044  0.0246     0.0  \n",
       "6665057  1.1072  0.0249     0.0  \n",
       "6665056  1.1399  0.0256     0.0  \n",
       "6665131  1.2958  0.0265     0.0  \n",
       "6665112  0.8220  0.0208     0.0  \n",
       "6665148  1.3511  0.0317     0.0  \n",
       "6665119  1.2585  0.0257     0.0  \n",
       "6665147  1.2895  0.0283     0.0  \n",
       "6665146  1.3043  0.0278     0.0  \n",
       "6665121  1.3503  0.0272     0.0  \n",
       "6665142  1.2498  0.0284     0.0  \n",
       "6665138  1.2736  0.0302     0.0  \n",
       "6665125  0.8790  0.0228     0.0  \n",
       "6665133  0.9160  0.0222     0.0  \n",
       "6665077  0.9457  0.0212     0.0  \n",
       "6665076  1.0651  0.0231     0.0  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Filtering the igap snps \n",
    "igap = igap.sort_values(by = ['Pvalue'], ascending=True)\n",
    "top_10_igap_snps = igap.iloc[0:20,:]\n",
    "top_10_igap_snps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "35343df4",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\ProgramData\\Anaconda3\\lib\\site-packages\\pandas\\core\\indexing.py:1720: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  self._setitem_single_column(loc, value, pi)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Position</th>\n",
       "      <th>MarkerName</th>\n",
       "      <th>Effect_allele</th>\n",
       "      <th>Non_Effect_allele</th>\n",
       "      <th>Beta</th>\n",
       "      <th>SE</th>\n",
       "      <th>Pvalue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6665069</th>\n",
       "      <td>19</td>\n",
       "      <td>44891712</td>\n",
       "      <td>rs184017</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>0.9704</td>\n",
       "      <td>0.0208</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6665078</th>\n",
       "      <td>19</td>\n",
       "      <td>44893408</td>\n",
       "      <td>rs59007384</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>0.9812</td>\n",
       "      <td>0.0208</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6665060</th>\n",
       "      <td>19</td>\n",
       "      <td>44885243</td>\n",
       "      <td>rs283811</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.9726</td>\n",
       "      <td>0.0218</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6665154</th>\n",
       "      <td>19</td>\n",
       "      <td>44923868</td>\n",
       "      <td>rs111789331</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>1.3811</td>\n",
       "      <td>0.0321</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6665058</th>\n",
       "      <td>19</td>\n",
       "      <td>44884873</td>\n",
       "      <td>rs34342646</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>1.1044</td>\n",
       "      <td>0.0246</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6665057</th>\n",
       "      <td>19</td>\n",
       "      <td>44884339</td>\n",
       "      <td>rs12972970</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>1.1072</td>\n",
       "      <td>0.0249</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6665056</th>\n",
       "      <td>19</td>\n",
       "      <td>44884202</td>\n",
       "      <td>rs12972156</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>1.1399</td>\n",
       "      <td>0.0256</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6665131</th>\n",
       "      <td>19</td>\n",
       "      <td>44912456</td>\n",
       "      <td>rs10414043</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>1.2958</td>\n",
       "      <td>0.0265</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6665112</th>\n",
       "      <td>19</td>\n",
       "      <td>44903416</td>\n",
       "      <td>rs10119</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.8220</td>\n",
       "      <td>0.0208</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6665148</th>\n",
       "      <td>19</td>\n",
       "      <td>44919689</td>\n",
       "      <td>rs41377151</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1.3511</td>\n",
       "      <td>0.0317</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6665119</th>\n",
       "      <td>19</td>\n",
       "      <td>44906745</td>\n",
       "      <td>rs769449</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>1.2585</td>\n",
       "      <td>0.0257</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6665147</th>\n",
       "      <td>19</td>\n",
       "      <td>44919589</td>\n",
       "      <td>rs56131196</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>1.2895</td>\n",
       "      <td>0.0283</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6665146</th>\n",
       "      <td>19</td>\n",
       "      <td>44918903</td>\n",
       "      <td>rs12721051</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>1.3043</td>\n",
       "      <td>0.0278</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6665121</th>\n",
       "      <td>19</td>\n",
       "      <td>44908684</td>\n",
       "      <td>rs429358</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1.3503</td>\n",
       "      <td>0.0272</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6665142</th>\n",
       "      <td>19</td>\n",
       "      <td>44917997</td>\n",
       "      <td>rs12721046</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>1.2498</td>\n",
       "      <td>0.0284</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6665138</th>\n",
       "      <td>19</td>\n",
       "      <td>44915533</td>\n",
       "      <td>rs5117</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1.2736</td>\n",
       "      <td>0.0302</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6665125</th>\n",
       "      <td>19</td>\n",
       "      <td>44910319</td>\n",
       "      <td>rs75627662</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.8790</td>\n",
       "      <td>0.0228</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6665133</th>\n",
       "      <td>19</td>\n",
       "      <td>44912921</td>\n",
       "      <td>rs483082</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>0.9160</td>\n",
       "      <td>0.0222</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6665077</th>\n",
       "      <td>19</td>\n",
       "      <td>44892962</td>\n",
       "      <td>rs157582</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.9457</td>\n",
       "      <td>0.0212</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6665076</th>\n",
       "      <td>19</td>\n",
       "      <td>44892887</td>\n",
       "      <td>rs11556505</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>1.0651</td>\n",
       "      <td>0.0231</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         Chromosome  Position   MarkerName Effect_allele Non_Effect_allele  \\\n",
       "6665069          19  44891712     rs184017             G                 T   \n",
       "6665078          19  44893408   rs59007384             T                 G   \n",
       "6665060          19  44885243     rs283811             G                 A   \n",
       "6665154          19  44923868  rs111789331             A                 T   \n",
       "6665058          19  44884873   rs34342646             A                 G   \n",
       "6665057          19  44884339   rs12972970             A                 G   \n",
       "6665056          19  44884202   rs12972156             G                 C   \n",
       "6665131          19  44912456   rs10414043             A                 G   \n",
       "6665112          19  44903416      rs10119             A                 G   \n",
       "6665148          19  44919689   rs41377151             G                 A   \n",
       "6665119          19  44906745     rs769449             A                 G   \n",
       "6665147          19  44919589   rs56131196             A                 G   \n",
       "6665146          19  44918903   rs12721051             G                 C   \n",
       "6665121          19  44908684     rs429358             C                 T   \n",
       "6665142          19  44917997   rs12721046             A                 G   \n",
       "6665138          19  44915533       rs5117             C                 T   \n",
       "6665125          19  44910319   rs75627662             T                 C   \n",
       "6665133          19  44912921     rs483082             T                 G   \n",
       "6665077          19  44892962     rs157582             T                 C   \n",
       "6665076          19  44892887   rs11556505             T                 C   \n",
       "\n",
       "           Beta      SE  Pvalue  \n",
       "6665069  0.9704  0.0208     0.0  \n",
       "6665078  0.9812  0.0208     0.0  \n",
       "6665060  0.9726  0.0218     0.0  \n",
       "6665154  1.3811  0.0321     0.0  \n",
       "6665058  1.1044  0.0246     0.0  \n",
       "6665057  1.1072  0.0249     0.0  \n",
       "6665056  1.1399  0.0256     0.0  \n",
       "6665131  1.2958  0.0265     0.0  \n",
       "6665112  0.8220  0.0208     0.0  \n",
       "6665148  1.3511  0.0317     0.0  \n",
       "6665119  1.2585  0.0257     0.0  \n",
       "6665147  1.2895  0.0283     0.0  \n",
       "6665146  1.3043  0.0278     0.0  \n",
       "6665121  1.3503  0.0272     0.0  \n",
       "6665142  1.2498  0.0284     0.0  \n",
       "6665138  1.2736  0.0302     0.0  \n",
       "6665125  0.8790  0.0228     0.0  \n",
       "6665133  0.9160  0.0222     0.0  \n",
       "6665077  0.9457  0.0212     0.0  \n",
       "6665076  1.0651  0.0231     0.0  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "for index,row in top_10_igap_snps.iterrows():\n",
    "    response = Entrez.efetch(db='SNP', id=str(top_10_igap_snps.loc[index,'MarkerName'])).read()\n",
    "    response = response[:-1]\n",
    "    response_o = xmltodict.parse(response)\n",
    "    pos = response_o['DocumentSummary']['CHRPOS']\n",
    "    pos = pos.split(':')[1]\n",
    "    top_10_igap_snps.loc[index,'Position'] = int(pos)\n",
    "\n",
    "top_10_igap_snps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "273edbc1",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Defining a SNP class to perform simple LD filtering duties\n",
    "class SNP:\n",
    "    \n",
    "    def __init__(self,rsid,position,chromosome):\n",
    "        self.rsid = rsid\n",
    "        self.position = position\n",
    "        self.chr = chromosome\n",
    "    \n",
    "\n",
    "        \n",
    "    def check_ld_snps(self,dataset,window = 1000):\n",
    "        start_position = self.position - window + 1\n",
    "        end_position = self.position + window\n",
    "        dataset = dataset[dataset['Chromosome'] == self.chr]\n",
    "        def extract_neighbour_snps(start_position, end_position, dataset):\n",
    "            neighbour_snps = []\n",
    "            for index,row in dataset.iterrows():\n",
    "                if start_position <= dataset.loc[index,'Position'] <= end_position:\n",
    "                    neighbour_snps.append(dataset.loc[index,'MarkerName'])\n",
    "                else:\n",
    "                    continue\n",
    "            return neighbour_snps\n",
    "    \n",
    "        self.snps_in_window = extract_neighbour_snps(start_position,end_position,dataset)\n",
    "        return self.snps_in_window\n",
    "    \n",
    "    def obtain_snp_sequence(self,window = 1000):\n",
    "        start_position = self.position - window +1\n",
    "        end_position = self.position + window\n",
    "        if int(self.chr) < 10:\n",
    "            id_chr = \"\".join([\"NC_00000\",str(self.chr)])\n",
    "        else:\n",
    "            id_chr = \"\".join([\"NC_0000\",str(self.chr)])\n",
    "\n",
    "        handle = Entrez.efetch(db=\"nucleotide\",\n",
    "                        id = id_chr,\n",
    "                        rettype = \"fasta\",\n",
    "                        strand = 1,\n",
    "                        seq_start = start_position,\n",
    "                        seq_stop  = end_position)\n",
    "        record = SeqIO.read(handle,\"fasta\")\n",
    "        self.snp_sequence = str(record.seq)\n",
    "        return self.snp_sequence\n",
    "    \n",
    "    def obtain_all_comb_seq(self,dataset,sign_num = 'null', window = 1000):\n",
    "        \n",
    "        def all_snp_combinations(a):\n",
    "            combinations = []\n",
    "            for k in range(0,len(a)):\n",
    "                t = list(itertools.combinations(a,k+1))\n",
    "                combinations.extend(t)\n",
    "            return combinations\n",
    "        \n",
    "        self.combinations = all_snp_combinations(self.snps_in_window)\n",
    "        comb_names = ['_'.join(x) for x in self.combinations if len(x)> 0]\n",
    "        comb_names.append('_'.join(['Ref',self.rsid]))\n",
    "        combination_dataset = dataset[dataset['MarkerName'].isin(self.snps_in_window)]\n",
    "        if sign_num != 'null':\n",
    "            combination_dataset = combination_dataset.sort_values('Pvalue')\n",
    "            combination_dataset = combination_dataset.iloc[0:int(sign_num),:]\n",
    "        sequences = []\n",
    "        \n",
    "        for comb in self.combinations:\n",
    "            seq_to_change = self.snp_sequence\n",
    "            start_position = self.position - window + 1\n",
    "            end_position = self.position + window\n",
    "            for k in range(0,len(comb)):\n",
    "                idx = combination_dataset['MarkerName'] == comb[k]\n",
    "                pos = combination_dataset.loc[idx,'Position']\n",
    "                allele = str(combination_dataset.loc[idx,'Non_Effect_allele'].values[0])\n",
    "                net_pos = int(pos) - int(start_position)\n",
    "                seq_to_change = seq_to_change[:net_pos-1] + allele + seq_to_change[net_pos:]\n",
    "            sequences.append(seq_to_change)\n",
    "        sequences.append(self.snp_sequence)\n",
    "        sequences_named = dict(zip(comb_names,sequences))\n",
    "        return sequences_named\n",
    "                \n",
    "                \n",
    "    def seq_combination(self,dataset,sign_num = 'null',window = 1000):\n",
    "        self.check_ld_snps(dataset,window)\n",
    "        self.obtain_snp_sequence()\n",
    "        self.combination_seq = self.obtain_all_comb_seq(dataset,sign_num,window)\n",
    "        return self.combination_seq\n",
    "        \n",
    "    \n",
    "    def __str__(self):\n",
    "        return \"The SNP in object is \"+self.rsid\n",
    "        \n",
    "        \n",
    "        \n",
    "        \n",
    "        "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2f1dd7a2",
   "metadata": {},
   "source": [
    "## Remodelling Chromatin.py from Expecto"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "bd85b1ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Important library calls\n",
    "import argparse\n",
    "import math\n",
    "import torch\n",
    "from torch import nn\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import h5py"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "f46ba8b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Inputing the resources for Expect.py\n",
    "inputsize = 2000\n",
    "batchSize = 32\n",
    "maxshift = 800\n",
    "args_cuda = False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "fa2029c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "#DL model\n",
    "class LambdaBase(nn.Sequential):\n",
    "    def __init__(self, fn, *args):\n",
    "        super(LambdaBase, self).__init__(*args)\n",
    "        self.lambda_func = fn\n",
    "\n",
    "    def forward_prepare(self, input):\n",
    "        output = []\n",
    "        for module in self._modules.values():\n",
    "            output.append(module(input))\n",
    "        return output if output else input\n",
    "\n",
    "class Lambda(LambdaBase):\n",
    "    def forward(self, input):\n",
    "        return self.lambda_func(self.forward_prepare(input))\n",
    "\n",
    "class Beluga(nn.Module):\n",
    "    def __init__(self):\n",
    "        super(Beluga, self).__init__()\n",
    "        self.model = nn.Sequential(\n",
    "            nn.Sequential(\n",
    "                nn.Conv2d(4,320,(1, 8)),\n",
    "                nn.ReLU(),\n",
    "                nn.Conv2d(320,320,(1, 8)),\n",
    "                nn.ReLU(),\n",
    "                nn.Dropout(0.2),\n",
    "                nn.MaxPool2d((1, 4),(1, 4)),\n",
    "                nn.Conv2d(320,480,(1, 8)),\n",
    "                nn.ReLU(),\n",
    "                nn.Conv2d(480,480,(1, 8)),\n",
    "                nn.ReLU(),\n",
    "                nn.Dropout(0.2),\n",
    "                nn.MaxPool2d((1, 4),(1, 4)),\n",
    "                nn.Conv2d(480,640,(1, 8)),\n",
    "                nn.ReLU(),\n",
    "                nn.Conv2d(640,640,(1, 8)),\n",
    "                nn.ReLU(),\n",
    "            ),\n",
    "            nn.Sequential(\n",
    "                nn.Dropout(0.5),\n",
    "                Lambda(lambda x: x.view(x.size(0),-1)),\n",
    "                nn.Sequential(Lambda(lambda x: x.view(1,-1) if 1==len(x.size()) else x ),nn.Linear(67840,2003)),\n",
    "                nn.ReLU(),\n",
    "                nn.Sequential(Lambda(lambda x: x.view(1,-1) if 1==len(x.size()) else x ),nn.Linear(2003,2002)),\n",
    "            ),\n",
    "            nn.Sigmoid(),\n",
    "        )\n",
    "\n",
    "    def forward(self, x):\n",
    "        return self.model(x)\n",
    "\n",
    "\n",
    "\n",
    "def encodeSeqs(seqs, inputsize=2000):\n",
    "    \"\"\"Convert sequences to 0-1 encoding and truncate to the input size.\n",
    "    The output concatenates the forward and reverse complement sequence\n",
    "    encodings.\n",
    "    Args:\n",
    "        seqs: list of sequences (e.g. produced by fetchSeqs)\n",
    "        inputsize: the number of basepairs to encode in the output\n",
    "    Returns:\n",
    "        numpy array of dimension: (2 x number of sequence) x 4 x inputsize\n",
    "    2 x number of sequence because of the concatenation of forward and reverse\n",
    "    complement sequences.\n",
    "    \"\"\"\n",
    "    seqsnp = np.zeros((len(seqs), 4, inputsize), np.bool_)\n",
    "\n",
    "    mydict = {'A': np.asarray([1, 0, 0, 0]), 'G': np.asarray([0, 1, 0, 0]),\n",
    "            'C': np.asarray([0, 0, 1, 0]), 'T': np.asarray([0, 0, 0, 1]),\n",
    "            'N': np.asarray([0, 0, 0, 0]), 'H': np.asarray([0, 0, 0, 0]),\n",
    "            'a': np.asarray([1, 0, 0, 0]), 'g': np.asarray([0, 1, 0, 0]),\n",
    "            'c': np.asarray([0, 0, 1, 0]), 't': np.asarray([0, 0, 0, 1]),\n",
    "            'n': np.asarray([0, 0, 0, 0]), '-': np.asarray([0, 0, 0, 0])}\n",
    "\n",
    "    n = 0\n",
    "    for line in seqs:\n",
    "        cline = line[int(math.floor(((len(line) - inputsize) / 2.0))):int(math.floor(len(line) - (len(line) - inputsize) / 2.0))]\n",
    "        for i, c in enumerate(cline):\n",
    "            seqsnp[n, :, i] = mydict[c]\n",
    "        n = n + 1\n",
    "\n",
    "    # get the complementary sequences\n",
    "    dataflip = seqsnp[:, ::-1, ::-1]\n",
    "    seqsnp = np.concatenate([seqsnp, dataflip], axis=0)\n",
    "    return seqsnp\n",
    "\n",
    "def get_predicted_diff(snp_comb_seq,inputsize = 2000, batchSize = 32, maxshift = 800, args_cuda = False):\n",
    "    \"\"\"\n",
    "    Function to obtain all the predicted chromatin values for reference and alterante \n",
    "    and find the difference among them for further analysis.\n",
    "    Args:\n",
    "        snp_comb_seq: A dictionary of sequences as string object with A,T,G,C characters\n",
    "                        and keys corresponding to snps and combinations of snps with atleast\n",
    "                        one snp having 'Ref' in the key name to denote reference variant\n",
    "    Return:\n",
    "            A dictionary of matrix size 4000x2002 for the chromatin difference values for each \n",
    "            variant and combination except the reference\n",
    "    \"\"\"\n",
    "    refseqs = [seq for key, seq in snp_comb_seq.items() if 'ref' in key.lower()]\n",
    "    ref_encoded = encodeSeqs(refseqs, inputsize=inputsize).astype(np.float32)\n",
    "\n",
    "    ref_preds = []\n",
    "    for i in range(int(1 + (ref_encoded.shape[0]-1) / batchSize)):\n",
    "        input = torch.from_numpy(ref_encoded[int(i*batchSize):int((i+1)*batchSize),:,:]).unsqueeze(2)\n",
    "        if args_cuda:\n",
    "            input = input.cuda()\n",
    "        ref_preds.append(model.forward(input).cpu().detach().numpy().copy())\n",
    "    ref_preds = np.vstack(ref_preds)\n",
    "    \n",
    "    comb_diff_pred = {}\n",
    "    for comb_seq in snp_comb_seq.keys():\n",
    "\n",
    "        if('Ref' not in comb_seq):\n",
    "\n",
    "            altseqs = [snp_comb_seq[comb_seq]]\n",
    "            alt_encoded = encodeSeqs(altseqs, inputsize=inputsize).astype(np.float32)\n",
    "\n",
    "            alt_preds = []\n",
    "            for i in range(int(1 + (alt_encoded.shape[0]-1) / batchSize)):\n",
    "                input = torch.from_numpy(alt_encoded[int(i*batchSize):int((i+1)*batchSize),:,:]).unsqueeze(2)\n",
    "                if args_cuda:\n",
    "                    input = input.cuda()\n",
    "                alt_preds.append(model.forward(input).cpu().detach().numpy().copy())\n",
    "            alt_preds = np.vstack(alt_preds)\n",
    "\n",
    "            diff = np.log2(ref_preds/(1-ref_preds)) - np.log2(alt_preds/(1-alt_preds)) \n",
    "            comb_diff_pred[comb_seq] = diff\n",
    "    \n",
    "    \n",
    "    return comb_diff_pred"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "71e90b4f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Beluga(\n",
       "  (model): Sequential(\n",
       "    (0): Sequential(\n",
       "      (0): Conv2d(4, 320, kernel_size=(1, 8), stride=(1, 1))\n",
       "      (1): ReLU()\n",
       "      (2): Conv2d(320, 320, kernel_size=(1, 8), stride=(1, 1))\n",
       "      (3): ReLU()\n",
       "      (4): Dropout(p=0.2, inplace=False)\n",
       "      (5): MaxPool2d(kernel_size=(1, 4), stride=(1, 4), padding=0, dilation=1, ceil_mode=False)\n",
       "      (6): Conv2d(320, 480, kernel_size=(1, 8), stride=(1, 1))\n",
       "      (7): ReLU()\n",
       "      (8): Conv2d(480, 480, kernel_size=(1, 8), stride=(1, 1))\n",
       "      (9): ReLU()\n",
       "      (10): Dropout(p=0.2, inplace=False)\n",
       "      (11): MaxPool2d(kernel_size=(1, 4), stride=(1, 4), padding=0, dilation=1, ceil_mode=False)\n",
       "      (12): Conv2d(480, 640, kernel_size=(1, 8), stride=(1, 1))\n",
       "      (13): ReLU()\n",
       "      (14): Conv2d(640, 640, kernel_size=(1, 8), stride=(1, 1))\n",
       "      (15): ReLU()\n",
       "    )\n",
       "    (1): Sequential(\n",
       "      (0): Dropout(p=0.5, inplace=False)\n",
       "      (1): Lambda()\n",
       "      (2): Sequential(\n",
       "        (0): Lambda()\n",
       "        (1): Linear(in_features=67840, out_features=2003, bias=True)\n",
       "      )\n",
       "      (3): ReLU()\n",
       "      (4): Sequential(\n",
       "        (0): Lambda()\n",
       "        (1): Linear(in_features=2003, out_features=2002, bias=True)\n",
       "      )\n",
       "    )\n",
       "    (2): Sigmoid()\n",
       "  )\n",
       ")"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model = Beluga()\n",
    "model.load_state_dict(torch.load('deepsea.beluga.pth'))\n",
    "model.eval()\n",
    "#model.cuda()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "2291e519",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"\\ncomb_diff_pred = get_predicted_diff(snp_comb_seq)\\nf = h5py.File(snp_test +'.diff.h5', 'w')\\nkey_names = list(comb_diff_pred.keys())\\nfor i in key_names:\\n    f.create_dataset(i, data=comb_diff_pred[i])\\nf.close()\\n\""
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"\n",
    "comb_diff_pred = get_predicted_diff(snp_comb_seq)\n",
    "f = h5py.File(snp_test +'.diff.h5', 'w')\n",
    "key_names = list(comb_diff_pred.keys())\n",
    "for i in key_names:\n",
    "    f.create_dataset(i, data=comb_diff_pred[i])\n",
    "f.close()\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f1e58481",
   "metadata": {},
   "source": [
    "## Running chromatin prediction for top 10 SNPs and its combinations with LD blocks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "fa142a88",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Position</th>\n",
       "      <th>MarkerName</th>\n",
       "      <th>Effect_allele</th>\n",
       "      <th>Non_Effect_allele</th>\n",
       "      <th>Beta</th>\n",
       "      <th>SE</th>\n",
       "      <th>Pvalue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>6552873</td>\n",
       "      <td>19</td>\n",
       "      <td>252541</td>\n",
       "      <td>rs12460337</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.0098</td>\n",
       "      <td>0.0302</td>\n",
       "      <td>0.7444</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>6552874</td>\n",
       "      <td>19</td>\n",
       "      <td>253938</td>\n",
       "      <td>rs7247199</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>-0.0090</td>\n",
       "      <td>0.0216</td>\n",
       "      <td>0.6783</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>6552875</td>\n",
       "      <td>19</td>\n",
       "      <td>254304</td>\n",
       "      <td>rs3929173</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>-0.0099</td>\n",
       "      <td>0.0212</td>\n",
       "      <td>0.6414</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>6552876</td>\n",
       "      <td>19</td>\n",
       "      <td>254447</td>\n",
       "      <td>rs1975527</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0207</td>\n",
       "      <td>0.0241</td>\n",
       "      <td>0.3904</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>6552877</td>\n",
       "      <td>19</td>\n",
       "      <td>254448</td>\n",
       "      <td>rs1975528</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>-0.0203</td>\n",
       "      <td>0.0235</td>\n",
       "      <td>0.3878</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>151294</th>\n",
       "      <td>6704167</td>\n",
       "      <td>19</td>\n",
       "      <td>58588397</td>\n",
       "      <td>rs149705456</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0049</td>\n",
       "      <td>0.0184</td>\n",
       "      <td>0.7880</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>151295</th>\n",
       "      <td>6704168</td>\n",
       "      <td>19</td>\n",
       "      <td>58590603</td>\n",
       "      <td>rs111790549</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>-0.0244</td>\n",
       "      <td>0.0455</td>\n",
       "      <td>0.5915</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>151296</th>\n",
       "      <td>6704169</td>\n",
       "      <td>19</td>\n",
       "      <td>58591012</td>\n",
       "      <td>rs79349307</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>-0.0171</td>\n",
       "      <td>0.0227</td>\n",
       "      <td>0.4507</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>151297</th>\n",
       "      <td>6704170</td>\n",
       "      <td>19</td>\n",
       "      <td>190188248</td>\n",
       "      <td>rs71223699</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>-0.0502</td>\n",
       "      <td>0.0556</td>\n",
       "      <td>0.3668</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>151298</th>\n",
       "      <td>6704171</td>\n",
       "      <td>19</td>\n",
       "      <td>58595163</td>\n",
       "      <td>rs148717940</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>-0.0404</td>\n",
       "      <td>0.0519</td>\n",
       "      <td>0.4367</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>151299 rows Ã— 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Unnamed: 0  Chromosome   Position   MarkerName Effect_allele  \\\n",
       "0          6552873          19     252541   rs12460337             A   \n",
       "1          6552874          19     253938    rs7247199             A   \n",
       "2          6552875          19     254304    rs3929173             A   \n",
       "3          6552876          19     254447    rs1975527             G   \n",
       "4          6552877          19     254448    rs1975528             T   \n",
       "...            ...         ...        ...          ...           ...   \n",
       "151294     6704167          19   58588397  rs149705456             G   \n",
       "151295     6704168          19   58590603  rs111790549             A   \n",
       "151296     6704169          19   58591012   rs79349307             G   \n",
       "151297     6704170          19  190188248   rs71223699             G   \n",
       "151298     6704171          19   58595163  rs148717940             T   \n",
       "\n",
       "       Non_Effect_allele    Beta      SE  Pvalue  \n",
       "0                      G  0.0098  0.0302  0.7444  \n",
       "1                      G -0.0090  0.0216  0.6783  \n",
       "2                      G -0.0099  0.0212  0.6414  \n",
       "3                      A  0.0207  0.0241  0.3904  \n",
       "4                      C -0.0203  0.0235  0.3878  \n",
       "...                  ...     ...     ...     ...  \n",
       "151294                 A  0.0049  0.0184  0.7880  \n",
       "151295                 G -0.0244  0.0455  0.5915  \n",
       "151296                 A -0.0171  0.0227  0.4507  \n",
       "151297                 A -0.0502  0.0556  0.3668  \n",
       "151298                 C -0.0404  0.0519  0.4367  \n",
       "\n",
       "[151299 rows x 9 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "skip_snps = ['rs59007384','rs111789331']\n",
    "apoe_snp = ['rs429358']\n",
    "igap_19 = pd.read_csv('IGAP_Chr19_Hg38loc.csv')\n",
    "igap_19"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "4a909a6d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running rs184017 ....\n",
      "Obtaining combinations for rs184017 ....\n",
      "Predicting the sequence profiles for rs184017 ....\n",
      "Saving the result for rs184017\n",
      "Running rs283811 ....\n",
      "Obtaining combinations for rs283811 ....\n",
      "Predicting the sequence profiles for rs283811 ....\n",
      "Saving the result for rs283811\n",
      "Running rs34342646 ....\n",
      "Obtaining combinations for rs34342646 ....\n",
      "Predicting the sequence profiles for rs34342646 ....\n",
      "Saving the result for rs34342646\n",
      "Running rs12972970 ....\n",
      "Obtaining combinations for rs12972970 ....\n",
      "Predicting the sequence profiles for rs12972970 ....\n",
      "Saving the result for rs12972970\n",
      "Running rs12972156 ....\n",
      "Obtaining combinations for rs12972156 ....\n",
      "Predicting the sequence profiles for rs12972156 ....\n",
      "Saving the result for rs12972156\n",
      "Running rs10414043 ....\n",
      "Obtaining combinations for rs10414043 ....\n",
      "Predicting the sequence profiles for rs10414043 ....\n",
      "Saving the result for rs10414043\n",
      "Running rs10119 ....\n",
      "Obtaining combinations for rs10119 ....\n",
      "Predicting the sequence profiles for rs10119 ....\n",
      "Saving the result for rs10119\n",
      "Running rs41377151 ....\n",
      "Obtaining combinations for rs41377151 ....\n",
      "Predicting the sequence profiles for rs41377151 ....\n",
      "Saving the result for rs41377151\n",
      "Running rs769449 ....\n",
      "Obtaining combinations for rs769449 ....\n",
      "Predicting the sequence profiles for rs769449 ....\n",
      "Saving the result for rs769449\n",
      "Running rs56131196 ....\n",
      "Obtaining combinations for rs56131196 ....\n",
      "Predicting the sequence profiles for rs56131196 ....\n",
      "Saving the result for rs56131196\n",
      "Running rs12721051 ....\n",
      "Obtaining combinations for rs12721051 ....\n",
      "Predicting the sequence profiles for rs12721051 ....\n",
      "Saving the result for rs12721051\n",
      "Running rs429358 ....\n",
      "Obtaining combinations for rs429358 ....\n",
      "Predicting the sequence profiles for rs429358 ....\n",
      "Saving the result for rs429358\n",
      "Running rs12721046 ....\n",
      "Obtaining combinations for rs12721046 ....\n",
      "Predicting the sequence profiles for rs12721046 ....\n",
      "Saving the result for rs12721046\n",
      "Running rs5117 ....\n",
      "Obtaining combinations for rs5117 ....\n",
      "Predicting the sequence profiles for rs5117 ....\n",
      "Saving the result for rs5117\n",
      "Running rs75627662 ....\n",
      "Obtaining combinations for rs75627662 ....\n",
      "Predicting the sequence profiles for rs75627662 ....\n",
      "Saving the result for rs75627662\n",
      "Running rs483082 ....\n",
      "Obtaining combinations for rs483082 ....\n",
      "Predicting the sequence profiles for rs483082 ....\n",
      "Saving the result for rs483082\n",
      "Running rs157582 ....\n",
      "Obtaining combinations for rs157582 ....\n",
      "Predicting the sequence profiles for rs157582 ....\n",
      "Saving the result for rs157582\n",
      "Running rs11556505 ....\n",
      "Obtaining combinations for rs11556505 ....\n",
      "Predicting the sequence profiles for rs11556505 ....\n",
      "Saving the result for rs11556505\n"
     ]
    }
   ],
   "source": [
    "\n",
    "for k in range(0,len(top_10_igap_snps)):\n",
    "    snp_test = top_10_igap_snps.iloc[k,2]\n",
    "    if snp_test not in skip_snps:\n",
    "        print(\"Running %s ....\"%(snp_test))\n",
    "        snp_obj = SNP(top_10_igap_snps.iloc[k,2],top_10_igap_snps.iloc[k,1],top_10_igap_snps.iloc[k,0])\n",
    "        print(\"Obtaining combinations for %s ....\"%(snp_test))\n",
    "        snp_comb_seq = snp_obj.seq_combination(igap_19)\n",
    "        print(\"Predicting the sequence profiles for %s ....\"%(snp_test))\n",
    "        comb_diff_pred = get_predicted_diff(snp_comb_seq)\n",
    "        f = h5py.File(snp_test +'.diff.h5', 'w')\n",
    "        key_names = list(comb_diff_pred.keys())\n",
    "        print(\"Saving the result for %s\"%(snp_test))\n",
    "        for i in key_names:\n",
    "            f.create_dataset(i, data=comb_diff_pred[i])\n",
    "        f.close()  \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a5f067a5",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "23d3961c",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
